### DISCUSSION TITLE
Why cuStateVec via Qsim with gpu_mode=(0,) slower than one with gpu_mode=1 for QFT circuit?

### DISCUSSION THREAD

**Discussion by rht** [NVIDIA/cuQuantum#35] [Q&A] [answered]
I was exploring the `gpu_mode` option on a NVIDIA cuQuantum Appliance Docker instance.
I was testing the usual QFT circuit on QsimCirq, and I tried the 2 options
```python
options = qsimcirq.QSimOptions(gpu_mode=(0,))
options = qsimcirq.QSimOptions(gpu_mode=1)
simulator = qsimcirq.QSimSimulator(options)
```

I found that `gpu_mode=(0,)` (multi-GPU backend but with only 1 GPU) to be consistently faster than the single-GPU backend.
My measurements for the elapsed time it takes to run the circuit, in seconds:
- QFT 28 qubits, `gpu_mode=(0,)`: 1.10,1.11,1.09
- QFT 28 qubits, `gpu_mode=1`: 1.45,1.45,1.45
- QFT 30 qubits, `gpu_mode=(0,)`: 3.67,3.80,3.59
- QFT 30 qubits, `gpu_mode=1`: 5.32,5.58,5.33

Why is this the case? Should I stay away from `gpu_mode=1`?

**Comment by leofang** [comment, marked as answer]
Hi @rht, thanks for sharing your observation. It's likely because there're extra optimizations that we offer with our custom mgpu backend, such as gate fusion. If you can use our cuQuantum Appliance container, it'd be better indeed if you use the mgpu backend even if you are on single GPU. But, for performance chasing games like this, it's always advised to benchmark with your workload first 🙂

**Reply by mtjrider** [reply]
As @leofang points out: in the appliance, `gpu_mode=(0,)` triggers the mulit-GPU backend running on a single GPU.
`gpu_mode=1` will force the simulator to use only cuStateVec as a backend.

The multi-GPU backend contains additional optimizations for circuit-level simulation.

[Here's the documentation link for reference](https://docs.nvidia.com/cuda/cuquantum/appliance/cirq.html#qsimcirq-qsimoptions).

**Reply by rht** [reply]
I see, my question is resolved. And I'm using cuStateVec for both options. I confirm that setting `max_fused_diagonal_gate_size=0` for `gpu_mode=(0,)` gives about the same elapsed time as `gpu_mode=1`. An optional question is that why is the diagonal gate fusion not applied in the single GPU backend?
